### PR TITLE
Bug 1908397 - Suggest metrics

### PR DESCRIPTION
**Pull request by bendk**
Defined metrics for suggest.  I used labeled timing distributions, which is a relatively new metric that seems great for our needs.  Using these metrics, we can track a separate timing distribution for each record type / provider.

Updated `ingest` to return metrics.  Added `query_with_metrics`, which returns metrics alongside the usual results. Added some classes to record metrics as the operations execute.

Moved the top-level `fetch_suggestions` from db.rs to store.rs.  It's now responsible for metrics, so I thought it fit better in store.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

**Comment by bendk** [comment]
I updated the CLI to report these metrics if the `-v` flag is present.  The results are interesting and feel like they're accurate to me:

```
========================== Ingestion times  ==========================
data                              Download:  1128ms Ingestion:  1488ms
fakespot-suggestions              Download:   321ms Ingestion:   687ms
amp-mobile-suggestions            Download:   190ms Ingestion:   655ms
pocket-suggestions                Download:    60ms Ingestion:    33ms
yelp-suggestions                  Download:    40ms Ingestion:    24ms
amo-suggestions                   Download:    36ms Ingestion:    13ms
mdn-suggestions                   Download:    34ms Ingestion:    11ms
weather                           Download:    19ms Ingestion:    10ms
configuration                     Download:    16ms Ingestion:     0ms
icon                              Download:   123ms Ingestion:     0ms
```

Query for "shoes" against all providers
```
============================ Query times  ============================
fakespot                          Time:  1551us
ampmobile                         Time:   362us
amp                               Time:   133us
pocket                            Time:   129us
weather                           Time:   116us
mdn                               Time:   105us
yelp                              Time:    91us
wikipedia                         Time:    56us
amo                               Time:    34us
```

**Comment by bendk** [comment]
The issue mentioned tracking total times in addition to per-provider/request-type times.  As I worked through this one though, it seemed like it would be better to do that in the dashboard.  We can just sum up the individual times to create the total.  Does that makes sense?

File: components/suggest/src/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+impl SuggestQueryMetrics {
+    pub fn measure_query<F, T>(&mut self, provider: impl Into<String>, operation: F) -> T
+    where
+        F: FnOnce() -> T,
+    {
+        let provider = provider.into();
+        let timer = Instant::now();
+        let result = operation();
```
Comment (ncloudioj) [inline components/suggest/src/metrics.rs:90]: The compiler is free to do whatever it likes to optimize the code: inlining, reordering, eliminating temporaries, etc. `std::hint::black_box()` is made to make the compiler pessimistic on that, it's often used for benchmarking. Compared to other metrics, query durations are more sensitive to those optimizations. Maybe we can wrap the operation inside the black box to ensure we're measuring the right thing.

**Review by ncloudioj — APPROVED** [review]
LGTM! This is great, thanks for making this happen!

File: components/suggest/src/metrics.rs (hunk 1 above)
Comment (bendk) [reply components/suggest/src/metrics.rs]: Great point, I'll update this.

**Pull request merged**